RHIDP-4807: Third-party plugin installation in RHDH (#733)
redhat-developer/red-hat-developers-documentation-rhdh: modules/dynamic-plugins/proc-export-third-party-plugins-rhdh.adoc

[id="proc-export-third-party-plugins-rhdh_{context}"]
= Exporting third-party plugins in {product}

To use plugins in {product}, the plugins must be exported as derived dynamic plugin package. These packages contain the plugin code and dependencies, ready for dynamic plugin integration into {product-short}.

> The third-party plugins include:

Reviewer: ```suggestion
```

Backend plugins::
+
--

> To ensure compatibility with the dynamic plugin support and enable their use as dynamic plugins, existing backend plugins must be compatible with the plugin instance. Additionally, these plugins must be rebuilt using a dedicated CLI command.

Reviewer: ```suggestion
To ensure compatibility with the dynamic plugin support and enable their use as dynamic plugins, existing backend plugins must be compatible with the new Backstage backend system. Additionally, these plugins must be rebuilt using a dedicated CLI command.
```

Author: To avoid using Backstage, I've changed the term to "new backend system"

> The plugin instance is created using `createBackendPlugin()` or `createBackendModule()`. You must export the plugin instance as the default export from either the main package or an `alpha` package (if the plugin instance support is still provided using `alpha` APIs). This doesn't add any additional requirement on top of the standard plugin development guidelines of the plugin instance.

Reviewer: ```suggestion
The new backend system entry point (created using `createBackendPlugin()` or `createBackendModule()`)  must be exported as the default export from either the main package or an `alpha` package (if the plugin instance support is still provided using `alpha` APIs). This doesn't add any additional requirement on top of the standard plugin development guidelines of the plugin instance.
```

> 

Reviewer: The following paragraph is missing from here. It is important as it gives a high-level description of how his process works. 

```
The dynamic export mechanism identifies private, non-backstage dependencies, and sets the `bundleDependencies` field in the `package.json` file for them, so that the dynamic plugin package can be published as a self-contained package, along with its private dependencies bundled in a private `node_modules` folder.
```

Author: Do we need to mention non-backstage? 
I've added the following info:

`The dynamic export mechanism identifies private dependencies and sets the `bundleDependencies` field in the `package.json` file. This export mechanism ensures that the dynamic plugin package is published as a self-contained package, with its private dependencies bundled in a private `node_modules` folder.`

Certain plugin dependencies require specific handling in the derived packages, such as:

* *Shared dependencies* are provided by the {product-very-short} application and listed as `peerDependencies` in `package.json` file, not bundled in the dynamic plugin package. For example, by default, all `@backstage` scoped packages are shared.
+
You can use the `--shared-package` flag to specify shared dependencies. To treat a `@backstage` package as private, use the negation prefix (`!`).

* *Embedded dependencies* are bundled into the dynamic plugin package with their dependencies hoisted to the top level. By default, packages with `-node` or `-common` suffixes are embedded.
+

> You can use the `--embed-package` flag to specify additional embedded packages, including those in the same repository that do not follow the default naming convention.

Reviewer: ```suggestion
You can use the `--embed-package` flag to specify additional embedded packages. For example packages from the same workspace that do not follow the default naming convention.
```

    "PluginRoot": "./src/index.ts"  // The default module name is "PluginRoot" and doesn't need explicit specification in the app-config.yaml file.
  }
}
----

You can add a `scalprum` section to the `package.json` file. For example:

.Example `scalprum` customization
[source,json]
----
"scalprum": {
  "name": "custom-package-name",
  "exposedModules": {
    "FooModuleName": "./src/foo.ts",
    "BarModuleName": "./src/bar.ts"
    // Define multiple modules here, with each exposed as a separate entry point in the Webpack container.
  }
}
----

Dynamic plugins might need adjustments for {product-short} needs, such as static JSX for mountpoints or dynamic routes. These changes are optional but might be incompatible with static plugins.

To include static JSX, define an additional export and use it as the dynamic plugin's `importName`. For example:

.Example static and dynamic plugin export
[source,tsx]
----
// For a static plugin
export const EntityTechdocsContent = () => {...}

// For a dynamic plugin
export const DynamicEntityTechdocsContent = {
  element: EntityTechdocsContent,
  staticJSXContent: (
    <TechDocsAddons>
      <ReportIssue />
    </TechDocsAddons>
  ),
};
----
--

> .Prerequisites

Reviewer: This (together with Procedure) should be at the begging of this document, right after the first introductory paragraph.  This is common for both frontend and backend

Author: The backend and frontend plugins specify information about the dependencies, which are also listed as prerequisites. I suggest we present the conceptual information about the dependencies first, followed by the prerequisites. This structure would make the procedure clearer for users as they follow the steps.
WDYT?

Reviewer 2: Consider splitting in multiple procedures:

*   _Exporting a third-party backend plugin_
*    _Exporting a third-party front-end plugin_

Consider also moving this Prequisistes section at the top of each procedure.

Then, consider clarifying lines 8 to 99 to make clear if these are very long prerequisites lists, or a first step "Prepare your plugin for exportation". (Seems more like a first step)

Reviewer: The current structure doesn't make sense. :-/ 
Everything below this line is common for both frontend and backend plugins and as such, I think it is better if it is at the beginning of this document as this will be enough for 90% of the use cases. After this generic explanation, we go deeper into the specifics of frontend and backend plugins.

* The `@janus-idp/cli` package is installed. Use the latest version (`@latest` tag) for compatibility with the most recent features and fixes.
* The third-party plugin must have a valid package.json file in its root directory, containing all required metadata and dependencies.
* Node.js and NPM is installed and configured.

> * You have access to a private NPM registry for publishing and distributing the exported plugin package.

Reviewer: ```suggestion
```
This is not a requirement for exporting

* The third-party plugin is compatible with your {product} version. For more information, see link:https://github.com/janus-idp/backstage-showcase/blob/main/docs/dynamic-plugins/versions.md[Version compatibility matrix].

.Procedure
Use the `package export-dynamic-plugin` command from the `@janus-idp/cli` package to export the plugin:

.Example command to export a third-party plugin
[source,bash]
----
npx @janus-idp/cli@latest package export-dynamic-plugin
----

Ensure that you execute the previous command in the root directory of the plugin's JavaScript package (containing `package.json` file).

> 

Reviewer: There should also be a warning about not pushing the derived package to npm registry

```
+
WARNING: The derived dynamic plugin JavaScript packages should not be pushed to the public npm registry. Generally, it is not necessary to push them to the npm registry. See xref:proc-package-third-party-dynamic-plugin_{context}[] for more suitable packaging optins. If you need to push it to npm registry, is your own private one.
+
```

[id="proc-export-third-party-plugins-rhdh_{context}"]
= Exporting third-party plugins in {product}

> To use plugins in {product}, the plugins must be exported as derived dynamic plugin package. These packages contain the plugin code and dependencies, ready for dynamic plugin integration into {product-short}.

Reviewer 2: It repeats the assembly introduction. Consider avoiding content repetition.

> Backend plugins::

Reviewer 2: Consider splitting in multiple procedures:

* _Exporting a third-party backend plugin_

* _Exporting a third-party front-end plugin_

Author: The procedure for exporting a backend or frontend plugin is the same; the only difference is the dependencies, which I’ve provided additional details about in separate sections. Therefore, I’ll leave it as is.

+
--
To ensure compatibility with the dynamic plugin support and enable their use as dynamic plugins, existing backend plugins must be compatible with the new backend system. Additionally, these plugins must be rebuilt using a dedicated CLI command.

> The new backend system entry point (created using `createBackendPlugin()` or `createBackendModule()`)  must be exported as the default export from either the main package or an `alpha` package (if the plugin instance support is still provided using `alpha` APIs). This doesn't add any additional requirement on top of the standard plugin development guidelines of the plugin instance.

Reviewer 3: ```suggestion
The backend system entry point (created using `createBackendPlugin()` or `createBackendModule()`)  must be exported as the default export from either the main package or an `alpha` package (if the plugin instance support is still provided using `alpha` APIs). This requirement aligns with the standard plugin development guidelines and does not introduce any additional rules for plugin instances.
```
I'm not sure we need to keep referring to the backend system as "new", unless something has changed in this release.

The other suggested change here is just cosmetic but I think it reads ok.

Author: @Gerry-Forde Good point.
@kadel Should we continue referring to it as the "new backend system," or would it be better to simply call it the "backend system"? By version 1.5, the "new" backend system may no longer be considered new, so this adjustment might be helpful for maintenance purposes.

Reviewer: It should be definitely a new backend system. This is how it is known by developers

The dynamic export mechanism identifies private dependencies and sets the `bundleDependencies` field in the `package.json` file. This export mechanism ensures that the dynamic plugin package is published as a self-contained package, with its private dependencies bundled in a private `node_modules` folder.

Certain plugin dependencies require specific handling in the derived packages, such as:

* *Shared dependencies* are provided by the {product-very-short} application and listed as `peerDependencies` in `package.json` file, not bundled in the dynamic plugin package. For example, by default, all `@backstage` scoped packages are shared.
+
You can use the `--shared-package` flag to specify shared dependencies, that are expected to be provided by {product} application and not bundled in the dynamic plugin package.
+
To treat a `@backstage` package as private, use the negation prefix (`!`). For example, when a plugin depends on the package in `@backstage` that is not provided by the {product} application. 

> * *Embedded dependencies* are bundled into the dynamic plugin package with their dependencies hoisted to the top level. By default, packages with `-node` or `-common` suffixes are embedded.

Reviewer 3: Is 'hoisted' the best term to use here...perhaps it is, I'm just unfamiliar with the phrase in this context :-)

+
You can use the `--embed-package` flag to specify additional embedded packages. For example, packages from the same workspace that do not follow the default naming convention.

The following is an example of exporting a dynamic plugin with shared and embedded packages:

.Example dynamic plugin export with shared and embedded packages
[source,bash]
----

> npx @janus-idp/cli@latest export-dynamic-plugin --shared-package '!/@backstage/plugin-notifications/'<1> --embed-package @backstage/plugin-notifications-backend <2>

Reviewer 3: In the GitHub preview the first label is not rendered correctly, maybe it needs to be escaped or a space added before '<1>'
<img width="817" alt="Screenshot 2024-12-11 at 12 03 04" src="https://github.com/user-attachments/assets/2f3550b4-cedb-4a4e-9e13-cc20cc1053bc">

The following is an example of exporting a dynamic plugin with shared and embedded packages:
+
.Example dynamic plugin export with shared and embedded packages

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```

--

Front-end plugins::
+
--

> Front-end plugins can use `scalprum` for configuration, which the CLI can generate automatically during the export process. The generated default configuration is logged when running the following command:

Reviewer 3: ```suggestion
Front-end plugins use `scalprum` for configuration, with the CLI automatically generating the configuration during the export process. The generated default configuration is logged when running the following command:
```

.Example command to log the default configuration
[source,bash]
----
npx @janus-idp/cli@latest export-dynamic
----

> The following is an example of default `scalprum` configuration:

Author: Here, the following is not a command but a JSON configuration.

> [source,bash]

Reviewer 2: ```suggestion
[source,terminal]
```

  }
}
----
+
You can add a `scalprum` section to the `package.json` file. For example:
+
.Example `scalprum` customization
[source,json]
----
"scalprum": {
  "name": "custom-package-name",
  "exposedModules": {
    "FooModuleName": "./src/foo.ts",
    "BarModuleName": "./src/bar.ts"
    // Define multiple modules here, with each exposed as a separate entry point in the Webpack container.
  }
}
----
+
Dynamic plugins might need adjustments for {product-short} needs, such as static JSX for mountpoints or dynamic routes. These changes are optional but might be incompatible with static plugins.
+
To include static JSX, define an additional export and use it as the dynamic plugin's `importName`. For example:
+
.Example static and dynamic plugin export
[source,tsx]
----
// For a static plugin
export const EntityTechdocsContent = () => {...}

// For a dynamic plugin
export const DynamicEntityTechdocsContent = {
  element: EntityTechdocsContent,
  staticJSXContent: (
    <TechDocsAddons>
      <ReportIssue />
    </TechDocsAddons>
  ),
};
----
--

> .Procedure

Reviewer: From this line to the end of this file, should be right after Prerequisites and before backend plugins

Author: @kadel According to the procedural module template, the module should conclude with a procedure. The details about backend and frontend plugins include some conceptual information, so I think adding them to the prerequisites section would help clarify these requirements for users. Therefore, we’ll retain them in the prerequisites.

.Procedure

> Use the `package export-dynamic-plugin` command from the `@janus-idp/cli` package to export the plugin:

Reviewer 3: Is this a single-step procedure, if so, I'd suggest making it a single item in an unordered list - see these examples: [Sample single-step procedures](https://docs.google.com/document/d/1e46KXZaSU2rxfZu6rP2bn3710RoTJpzeD2vlh8kRDG0/edit?tab=t.0)

.Example command to export a third-party plugin

> [source,bash]

Reviewer 2: This is not a bash script, but a terminal command:

```suggestion
[source,terminal]
```

Backend plugins::
+
--
To ensure compatibility with the dynamic plugin support and enable their use as dynamic plugins, existing backend plugins must be compatible with the backend system. Additionally, these plugins must be rebuilt using a dedicated CLI command.

> The backend system entry point (created using `createBackendPlugin()` or `createBackendModule()`)  must be exported as the default export from either the main package or an `alpha` package (if the plugin instance support is still provided using `alpha` APIs). This doesn't add any additional requirement on top of the standard plugin development guidelines of the plugin instance.

Author: Changed it to "new Backstage backend system"
